django/contrib/gis/db/backends/postgis/base.py

```diff
@@ -27,3 +136,59 @@ def prepare_database(self):
             if bool(cursor.fetchone()):
                 return
             cursor.execute("CREATE EXTENSION IF NOT EXISTS postgis")
+            self.register_geometry_adapters(self.connection)
+
+    def get_new_connection(self, conn_params):
+        connection = super().get_new_connection(conn_params)
+        self.register_geometry_adapters(connection)
+        return connection
+
+    if is_psycopg3:
+
+        def register_geometry_adapters(self, pg_connection):
+            from psycopg.types import TypeInfo
+
+            if self.alias not in self._geometry_types:
+                geo_info = TypeInfo.fetch(pg_connection, "geometry")
+                if not geo_info:
+                    return
+                self._geometry_types[self.alias] = geo_info
+
+            if self.alias not in self._geography_types:
+                geog_info = TypeInfo.fetch(pg_connection, "geography")
+                if not geog_info:
+                    return
+                self._geography_types[self.alias] = geog_info
+
+            if self.alias not in self._raster_types:
+                raster_info = TypeInfo.fetch(pg_connection, "raster")
+                if not raster_info:
+                    return
```

Comment:
Why short circuit with `return` rather than: `if raster_info: self._raster_types[self.alias] = raster_info`? On CockroachDB, there's no raster type so the method call unexpectedly ends here.

Reply:
Because it looks like I have no idea what what I am doing :) You are right, neither of those should return early. Will fix.